[2026-06 LWG Motion 6] P3395R6 (Fix encoding issues and add a formatter for std::error_code) (#9129)
cplusplus/draft: source/diagnostics.tex

\begin{itemdescr}
\pnum
\returns
A string of multibyte characters in the execution character set that describes the error condition

> denoted by \tcode{ev}.

Reviewer: ```suggestion
A string of multibyte characters in the execution character set
that describes the error condition denoted by \tcode{ev}.
```

Equivalent to: \tcode{return os << ec.category().name() << ':' << ec.value();}
\end{itemdescr}

> \rSec3[system.error.fmt]{Formatting}

Reviewer 2: It's not your fault, but this subclause stable name makes no sense. Everything else is called [syserr.*] except for [system.error.syn].

We probably want something like [syserr.fmt] or whatever the convention is for "formatting support" stable names.

\pnum
\begin{codeblock}

> template<class charT> struct formatter<error_code, charT> {

Reviewer 2: Missing indexing.

  constexpr void set_debug_format();

  constexpr typename basic_format_parse_context<charT>::iterator
    parse(basic_format_parse_context<charT>& ctx);

  template<class Out>
    typename basic_format_context<Out, charT>::iterator
      format(const error_code& ec, basic_format_context<Out, charT>& ctx) const;
};
\end{codeblock}

> \pnum

Reviewer 2: ```suggestion
```

\begin{itemdecl}
constexpr void set_debug_format();
\end{itemdecl}

\begin{itemdescr}
\pnum
\effects
Modifies the state of the \tcode{formatter} to be as if the

> \defnx{error-code-format-spec}{error-code-format-spec!syserr err fmt} passed by the

Reviewer 2: ```suggestion
\grammarterm{error-code-format-spec} passed by the
```
The paper italics are grammarterm italics, not defining.

last call to \tcode{parse} contained the \tcode{?} option.
\end{itemdescr}

> \pnum

Reviewer 2: ```suggestion
```

\begin{itemdecl}
constexpr typename basic_format_parse_context<charT>::iterator
  parse(basic_format_parse_context<charT>& ctx);
\end{itemdecl}

\begin{itemdescr}
\pnum
\effects
Parses the format specifier as a \defnx{error-code-format-spec}{error-code-format-spec!syserr err fmt}
and stores the parsed specifiers in \tcode{*this}.

\begin{ncbnf}
\fmtnontermdef{error-code-format-spec}\br
    \opt{fill-and-align} \opt{width} \opt{\tcode{?}} \opt{\tcode{s}}
\end{ncbnf}

> where the productions and \defnx{width}{width!format string} are described in\iref{format.string}

Reviewer 2: ```suggestion
where the productions \grammarterm{fill-and-align} and \grammarterm{width} are described in\iref{format.string}.
```

\pnum
\returns
An iterator past the end of the \defnx{error-code-format-spec}{error-code-format-spec!syserr err fmt}.
\end{itemdescr}

\pnum
\begin{itemdecl}
template<class Out>
  typename basic_format_context<Out, charT>::iterator
    format(const error_code& ec, basic_format_context<Out, charT>& ctx) const;
\end{itemdecl}

\begin{itemdescr}
\pnum
\effects
If the \tcode{s} option is used, then:
\begin{itemize}
\item If \tcode{charT} is \tcode{char} and the ordinary literal encoding is UTF-8, let
  \tcode{msg} be \tcode{ec.message()} transcoded to UTF-8 with maximal subparts of ill-formed
  subsequences substituted with \unicode{fffd}{replacement character} per the Unicode Standard, Chapter
  3.9 \unicode{fffd} Substitution in Conversion.

> \item Otherwise, let \tcode{msg} be \tcode{ec.message()} transcoded to an implementation-defined encoding.

Reviewer 2: Needs some reflowing:
```suggestion
\item
If \tcode{charT} is \tcode{char} and the ordinary literal encoding is UTF-8,
let \tcode{msg} be \tcode{ec.message()} transcoded to UTF-8
with maximal subparts of ill-formed subsequences substituted with \unicode{fffd}{replacement character}
per the Unicode Standard, Chapter 3.9 \unicode{fffd} Substitution in Conversion.
\item
Otherwise, let \tcode{msg} be \tcode{ec.message()}
transcoded to an implementation-defined encoding.
```

    parse(basic_format_parse_context<charT>& ctx);

  template<class Out>
    typename basic_format_context<Out, charT>::iterator
      format(const error_code& ec, basic_format_context<Out, charT>& ctx) const;
};
\end{codeblock}

\indexlibrarymember{set_debug_format}{formatter}%
\begin{itemdecl}
constexpr void set_debug_format();
\end{itemdecl}

\begin{itemdescr}
\pnum
\effects
Modifies the state of the \tcode{formatter} to be as if the
\fmtgrammarterm{error-code-format-spec} passed by the
last call to \tcode{parse} contained the \tcode{?} option.
\end{itemdescr}

\indexlibrarymember{parse}{formatter}%
\begin{itemdecl}
constexpr typename basic_format_parse_context<charT>::iterator
  parse(basic_format_parse_context<charT>& ctx);
\end{itemdecl}

\begin{itemdescr}
\pnum
\effects
Parses the format specifier as a \fmtgrammarterm{error-code-format-spec}
and stores the parsed specifiers in \tcode{*this}.

\begin{ncbnf}
\fmtnontermdef{error-code-format-spec}\br
    \opt{fill-and-align} \opt{width} \opt{\tcode{?}} \opt{\tcode{s}}
\end{ncbnf}

where the productions and \fmtgrammarterm{fill-and-align} and \fmtgrammarterm{width} are

> described in\iref{format.string}.

Reviewer 2: ```suggestion
described in \ref{format.string}.
```

\pnum
\returns
An iterator past the end of the \fmtgrammarterm{error-code-format-spec}.
\end{itemdescr}

\indexlibrarymember{format}{formatter}%
\begin{itemdecl}
template<class Out>
  typename basic_format_context<Out, charT>::iterator
    format(const error_code& ec, basic_format_context<Out, charT>& ctx) const;
\end{itemdecl}

\begin{itemdescr}
\pnum
\effects
If the \tcode{s} option is used, then:
\begin{itemize}
\item
If \tcode{charT} is \tcode{char} and the ordinary literal encoding is UTF-8,

> let \tcode{msg} be \tcode{ec.message()} transcoded to UTF-8

Reviewer 2: ```suggestion
then let \tcode{msg} be \tcode{ec.message()} transcoded to UTF-8
```

with maximal subparts of ill-formed subsequences substituted with \unicode{fffd}{replacement character}
per the Unicode Standard, Chapter 3.9 \unicode{fffd} Substitution in Conversion.
\item
Otherwise, let \tcode{msg} be \tcode{ec.message()}
transcoded to an implementation-defined encoding.
\end{itemize}

\pnum
Otherwise, let \tcode{msg} be \tcode{std::format("{}:{}", ec.category().name(), ec.value())}.

\pnum

> If the \tcode{?} option is used then \tcode{msg} is formatted as an escaped string(\iref{format.string.escaped}).

Reviewer 2: ```suggestion
If the \tcode{?} option is used
then \tcode{msg} is formatted as an escaped string\iref{format.string.escaped}.
```

> Writes \tcode{msg} into \tcode{ctx.out()}, adjusted according to the \fmtgrammarterm{error-code-format-spec}.

Reviewer 2: This all seems to be a scription of *Effects* to me, not separate paragraphs, but maybe I'm missing something. The paper does put gaps here, but idk, this is weird.
```suggestion
Otherwise, let \tcode{msg} be \tcode{std::format("{}:{}", ec.category().name(), ec.value())}.
If the \tcode{?} option is used then \tcode{msg} is formatted as an escaped string(\iref{format.string.escaped}).
Writes \tcode{msg} into \tcode{ctx.out()}, adjusted according to the \fmtgrammarterm{error-code-format-spec}.
```

Reviewer 3: I think we do have multi-paragraph \effects in some places, but having a separate _numbered_ paragraph for every sentence is certainly non-optimal.

Author: I have an idea

Author: <img width="698" height="212" alt="image" src="https://github.com/user-attachments/assets/614d97d2-23cd-4a48-b056-a5fb585a5b7d" />
I think this works?

Reviewer: I don't love the use of a list with no other text before or after the list. If we went this form I think the final sentence should be its own bullet. "Writes `msg` into `ctx.out()`" applies to all cases, not only when `?` is used.

We could just make it all one paragraph with no paragraph breaks, but keep the innermost list for the sub-clauses of the "If the `s` options is used" case. So exactly as in the incoming paper, except without a paragraph break before "If the `?` option is used".

Reviewer 4: That seems entirely adequate, thanks!

Equivalent to: \tcode{return os << ec.category().name() << ':' << ec.value();}
\end{itemdescr}

\rSec3[syserr.fmt]{Formatting}
\pnum
\indexlibraryglobal{formatter}%
\begin{codeblock}
template<class charT> struct formatter<error_code, charT> {
  constexpr void set_debug_format();

  constexpr typename basic_format_parse_context<charT>::iterator
    parse(basic_format_parse_context<charT>& ctx);

  template<class Out>
    typename basic_format_context<Out, charT>::iterator
      format(const error_code& ec, basic_format_context<Out, charT>& ctx) const;
};
\end{codeblock}

\indexlibrarymember{set_debug_format}{formatter}%
\begin{itemdecl}
constexpr void set_debug_format();
\end{itemdecl}

\begin{itemdescr}
\pnum
\effects
Modifies the state of the \tcode{formatter} to be as if the

> \fmtgrammarterm{error-code-format-spec} passed by the

Reviewer: ```suggestion
\fmtgrammarterm{error-code-format-spec} parsed by the
```

last call to \tcode{parse} contained the \tcode{?} option.
\end{itemdescr}

\indexlibrarymember{parse}{formatter}%
\begin{itemdecl}
constexpr typename basic_format_parse_context<charT>::iterator
  parse(basic_format_parse_context<charT>& ctx);
\end{itemdecl}

\begin{itemdescr}
\pnum
\effects

> Parses the format specifier as a \fmtgrammarterm{error-code-format-spec}

Reviewer: ```suggestion
Parses the format specifier as an \fmtgrammarterm{error-code-format-spec}
```

(this is wrong in the incoming paper, but we should fix it)

and stores the parsed specifiers in \tcode{*this}.

\begin{ncbnf}
\fmtnontermdef{error-code-format-spec}\br
    \opt{fill-and-align} \opt{width} \opt{\tcode{?}} \opt{\tcode{s}}
\end{ncbnf}

> where the productions and \fmtgrammarterm{fill-and-align} and \fmtgrammarterm{width} are

Reviewer: ```suggestion
where the productions \fmtgrammarterm{fill-and-align} and \fmtgrammarterm{width} are
```

described in \ref{format.string}.

\pnum
\returns
An iterator past the end of the \fmtgrammarterm{error-code-format-spec}.
\end{itemdescr}

\indexlibrarymember{format}{formatter}%
\begin{itemdecl}
template<class Out>
  typename basic_format_context<Out, charT>::iterator
    format(const error_code& ec, basic_format_context<Out, charT>& ctx) const;
\end{itemdecl}

\begin{itemdescr}
\pnum
\effects
\begin{itemize}
\item
If the \tcode{s} option is used, then:
\begin{itemize}
\item
If \tcode{charT} is \tcode{char} and the ordinary literal encoding is UTF-8,
then let \tcode{msg} be \tcode{ec.message()} transcoded to UTF-8
with maximal subparts of ill-formed subsequences substituted with \unicode{fffd}{replacement character}
per the Unicode Standard, Chapter 3.9 \unicode{fffd} Substi\-tution in Conversion.
\item
Otherwise, let \tcode{msg} be \tcode{ec.message()}
transcoded to an implementation-defined encoding.
\end{itemize}

\item

> Otherwise, let \tcode{msg} be \tcode{std::format("{}:{}", ec.category().name(), ec.value())}.

Reviewer: No `std::` here and you need to escape the `{}` replacement fields or LaTeX eats them.

\item
If the \tcode{?} option is used

> then \tcode{msg} is formatted as an escaped string(\iref{format.string.escaped}).

Reviewer: ```suggestion
then \tcode{msg} is formatted as an escaped string\iref{format.string.escaped}.
```
